src/Particle/VirtualParticleSet.h

```diff
@@ -0,0 +1,59 @@
+//////////////////////////////////////////////////////////////////////////////////////
+// This file is distributed under the University of Illinois/NCSA Open Source License.
+// See LICENSE file in top directory for details.
+//
+// Copyright (c) 2016 Jeongnim Kim and QMCPACK developers.
+//
+// File developed by: Ye Luo, yeluo@anl.gov, Argonne National Laboratory
+//                    Jeongnim Kim, jeongnim.kim@gmail.com, University of Illinois at Urbana-Champaign
+//
+// File created by: Jeongnim Kim, jeongnim.kim@gmail.com, University of Illinois at Urbana-Champaign
+//////////////////////////////////////////////////////////////////////////////////////
+
+
+/** @file VirtualParticleSet.h
+ * A proxy class to the quantum ParticleSet
+ */
+#ifndef QMCPLUSPLUS_VIRTUAL_PARTICLESET_H
+#define QMCPLUSPLUS_VIRTUAL_PARTICLESET_H
+
+#include <Utilities/Configuration.h>
+#include <Particle/ParticleSet.h>
+
+namespace qmcplusplus
+{
+/** Introduced to handle virtual moves and ratio computations, e.g. for non-local PP evaluations.
+   */
+class VirtualParticleSet : public ParticleSet
+{
+private:
+  /// true, if virtual particles are on a sphere for NLPP
+  bool onSphere;
+
+public:
+  /// Reference particle
+  int refPtcl;
+  /// Reference source particle, used when onSphere=true
+  int refSourcePtcl;
+
+  /// ParticleSet this object refers to
+  const ParticleSet& refPS;
+
+  inline bool isOnSphere() const { return onSphere; }
+
+  /** constructor 
+     * @param p ParticleSet whose virtual moves are handled by this object
+     * @param nptcl number of virtual particles
+     */
+  VirtualParticleSet(const ParticleSet& p, int nptcl);
+
```

Comment:
why not delete the unsafe copy constructor.

Reply:
The safety issue comes from ParticleSet. Not really VirtualParticleSet. ParticleSet hard-coded self-self distance to table[0]. This needs a bit cleaning effort. So once ParticleSet copy constructor is fixed, VirtualParticleSet default copy constructor should work. If I delete it, the code doesn't compile because std::vector::resize() needs it.

Reply:
I see there should be a comment on this in both repos

Reply:
Extra comments added. In QMCPACK repo, I opened an issue and planned at least two PRs to fix the issue. After everything is settled there, I will fix miniQMC. Currently "to be fixed" notes have been added.